service/src/main/java/bio/terra/catalog/service/DatasetService.java

```diff
@@ -132,11 +135,14 @@ private List<DatasetResponse> collectDatasets(StorageSystem system) {
     return createDatasetResponses(roleMap, system);
   }
 
-  public DatasetsListResponse listDatasets() {
+  public DatasetsListResponse listDatasets() throws InterruptedException {
     List<DatasetResponse> datasets = new ArrayList<>();
+    ExecutorService es = Executors.newCachedThreadPool();
     for (StorageSystem system : StorageSystem.values()) {
-      datasets.addAll(collectDatasets(system));
+      es.execute(() -> datasets.addAll(collectDatasets(system)));
```

Comment:
I'm surprised that this works. So the current user token, which is request scoped, is propagated to the other threads correctly?

Reply:
I am not actually sure. I am poking around a bit and we will see. Its definitely getting some results, but I am not sure what it is getting that is admin, rather than from the actual fetching. Going to comment out the admin code to see.

Reply:
Yeah, it doesn't work. I'll poke around.